### PR TITLE
[12.0][MIG] res_partner_operating_unit

### PR DESCRIPTION
**Pull request by Dangar-Serpentcs**

### CODE REVIEW

File: res_partner_operating_unit/models/res_partner.py [hunk 1]
```diff
@@ -0,0 +1,15 @@
+# © 2017 Niaga Solution - Edi Santoso <repodevs@gmail.com>
+# License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl.html).
+from odoo import fields, models
+
+
+class ResPartner(models.Model):
+
+    _inherit = 'res.partner'
+
+    operating_unit_id = fields.Many2one(
```
Comment (max3903) [inline res_partner_operating_unit/models/res_partner.py]: @Dangar-Serpentcs Can you change the type to a many2many field and adapt the record rules?

The default value should remain the default OU from the user.

File: res_partner_operating_unit/readme/DESCRIPTION.rst [hunk 2]
```diff
@@ -0,0 +1,7 @@
+This module introduces the following features:
+
+* Adds the Operating Unit (OU) to res partner.
+
```
Comment (max3903) [inline res_partner_operating_unit/readme/DESCRIPTION.rst]:
```suggestion
```

File: res_partner_operating_unit/readme/DESCRIPTION.rst [hunk 3]
```diff
@@ -0,0 +1,7 @@
...
+* The user’s default Operating Unit (OU) is proposed at the time of creating Partner.
+
```
Comment (max3903) [inline res_partner_operating_unit/readme/DESCRIPTION.rst]:
```suggestion
```

File: res_partner_operating_unit/models/res_partner.py (hunk 1 above)
Comment (Dangar-Serpentcs) [reply res_partner_operating_unit/models/res_partner.py]: Sure

**Comment by Dangar-Serpentcs** [comment]
Improved!

**Comment by Dangar-Serpentcs** [comment]
Improved!

File: res_partner_operating_unit/models/res_partner.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+        return self.operating_unit_default_get()
+
+    @api.model
+    def _default_operating_units(self):
+        return self._default_operating_unit()
+
+    operating_unit_ids = fields.Many2many(
+        'operating.unit', 'operating_unit_partner_rel', 'partner_id', 'poid',
+        'Operating Units', default=lambda self: self._default_operating_units()
+    )
```
Comment (nikul-serpentcs) [inline res_partner_operating_unit/models/res_partner.py]:
```suggestion
    )
 
```

File: res_partner_operating_unit/models/res_partner.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+    @api.model
+    def _default_operating_unit(self):
+        return self.operating_unit_default_get()
+
+    @api.model
+    def _default_operating_units(self):
+        return self._default_operating_unit()
+
+    operating_unit_ids = fields.Many2many(
+        'operating.unit', 'operating_unit_partner_rel', 'partner_id', 'poid',
```
Comment (max3903) [inline res_partner_operating_unit/models/res_partner.py]:
```suggestion
        'operating.unit', 'operating_unit_partner_rel', 'partner_id', 'operating_unit_id',
```

File: res_partner_operating_unit/tests/test_res_partner_operating_unit.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+        user = self.res_users_model.create({
+            'name': 'Test User',
+            'login': login,
+            'password': 'demo',
+            'email': 'test@yourcompany.com',
+            'company_id': company.id,
+            'company_ids': [(4, company.id)],
+            'operating_unit_ids': [(4, ou.id) for ou in operating_units],
+        })
+        return user
```
Comment (max3903) [inline res_partner_operating_unit/tests/test_res_partner_operating_unit.py:51]: @Dangar-Serpentcs You need to update `_create_partner` to set the many2many field and you need to add the tests to:

* check that `user` can access `partner`
* create a new partner in another OU and make sure `user` cannot access this new partner.

File: res_partner_operating_unit/views/res_partner_view.xml [hunk 7]
```diff
@@ -0,0 +1,41 @@
...
+    License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl-3.0) -->
+
+    <!-- Partner Form View  -->
+    <record model="ir.ui.view" id="view_partner_form">
+    <field name="name">res.partner.form</field>
+    <field name="model">res.partner</field>
+    <field name="inherit_id" ref="base.view_partner_form"/>
+    <field name="arch" type="xml">
+        <field name="category_id" position="after">
+            <field name="operating_unit_ids" groups="operating_unit.group_multi_operating_unit" domain = "[('company_id','=', company_id)]" options="{'no_create': True}"/>
```
Comment (max3903) [inline res_partner_operating_unit/views/res_partner_view.xml]:
```suggestion
            <field name="operating_unit_ids"
                   groups="operating_unit.group_multi_operating_unit"
                   domain="[('company_id', '=', company_id)]"
                   options="{'no_create': True}"
                   widget="many2many_tags"/>
```

**Review by nikul-serpentcs — APPROVED** [review]
Code Review LGTM :+1:

File: res_partner_operating_unit/views/res_partner_view.xml (hunk 7 above)
Comment (max3903) [reply res_partner_operating_unit/views/res_partner_view.xml]: @Dangar-Serpentcs You are missing the widget...

Comment (HirenDangar) [reply res_partner_operating_unit/views/res_partner_view.xml]: Improved

File: res_partner_operating_unit/views/res_partner_view.xml [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+    <!-- Partner Form View  -->
+    <record model="ir.ui.view" id="view_partner_form">
+    <field name="name">res.partner.form</field>
+    <field name="model">res.partner</field>
+    <field name="inherit_id" ref="base.view_partner_form"/>
+    <field name="arch" type="xml">
+        <field name="category_id" position="after">
+            <field name="operating_unit_ids"
+                   groups="operating_unit.group_multi_operating_unit"
+                   domain = "[('company_id','=', company_id)]"
```
Comment (max3903) [inline res_partner_operating_unit/views/res_partner_view.xml]:
```suggestion
                   domain="[('company_id','=', company_id)]"
```

**Review by max3903 — APPROVED** [review]

**Comment by max3903** [comment]
@Dangar-Serpentcs Please review and merge https://github.com/Dangar-Serpentcs/operating-unit/pull/4

File: res_partner_operating_unit/models/res_users.py [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+            self.partner_id.operating_unit_ids = \
+                [(4, res.default_operating_unit_id.id)]
+        return res
+
+    @api.constrains('partner_id.operating_unit_ids',
+                    'default_operating_unit_id')
+    def check_partner_operating_unit(self):
+        if self.partner_id.operating_unit_ids and \
+                self.default_operating_unit_id.id not in \
+                self.partner_id.operating_unit_ids:
```
Comment (max3903) [inline res_partner_operating_unit/models/res_users.py]:
```suggestion
                self.partner_id.operating_unit_ids.ids:
```

Comment (max3903) [reply res_partner_operating_unit/models/res_users.py]: @Dangar-Serpentcs This should fix the Travis errors.

**Comment by MiquelRForgeFlow** [comment]
@max3903 @Dangar-Serpentcs @nikul-serpentcs  It seems to me this module is wrong or incomplete. You can create inconsistencies like having a user with one operating unit and the partner of that user having a different operating unit.

Please, in order for OCA to continue having good quality, I request you to fix this.

**Pull request merged**